platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/isolated/IsolatedProjectsCompositeBuildIntegrationTest.groovy

```diff
@@ -47,4 +49,170 @@ class IsolatedProjectsCompositeBuildIntegrationTest extends AbstractIsolatedProj
         then:
         fixture.assertStateLoaded()
     }
+
+    def "cycles for plugin builds are prohibited"() {
+        given:
+        includePluginBuild(settingsFile, "plugins-a")
+        applyPlugins(buildFile, ["plugin-a"])
+
+        includedBuild("plugins-a") {
+            includePluginBuild(settingsScript, "../plugins-b")
+            applyPlugins(buildScript, ["groovy-gradle-plugin", "plugin-b"])
+            srcMainGroovy.file("plugin-a.gradle") << ""
+        }
+
+        includedBuild("plugins-b") {
+            includePluginBuild(settingsScript, "../plugins-c")
+            applyPlugins(buildScript, ["groovy-gradle-plugin", "plugin-c"])
+            srcMainGroovy.file("plugin-b.gradle") << ""
+        }
+
+        includedBuild("plugins-c") {
+            includePluginBuild(settingsScript, "../plugins-a")
+            applyPlugins(buildScript, ["groovy-gradle-plugin", "plugin-a"])
+            srcMainGroovy.file("plugin-c.gradle") << ""
+        }
+
+        when:
+        isolatedProjectsFails("help")
+
+        then:
+        failureDescriptionContains("A cycle has been detected in the definition of plugin builds: :plugins-c -> :plugins-a -> :plugins-b -> :plugins-c.")
```

Comment:
The whole trace doesn't make sense for cycles scenario imo. 
Consider the whole trace: `: -> :a -> :b -> :c -> d: -> :e -> :c`. To understand that `:c` brings the cycle, you have to iterate to the very end of the trace, and then start backtracking.

Effectively, the "tail" (or, a "head", depend on perspective, `: -> :a -> :b` in the example) of the whole path is doesn't matter: to fix the cycle you have to jump to "offender" node `:c`, and untie its edges.

The [next PR](https://github.com/gradle/gradle/pull/31070), stacked on the current, is improving the cycle detection logic, and guarantees, that cycle path will always be started from "acyclic" node, what is a `plugin build` node in case of build tree.

Reply:
> To understand that :c brings the cycle, you have to iterate to the very end of the trace, and then start backtracking.

Not necessarily. The cycle can be marked visually with brackets. Or the path to the cycle and the cycle can be printed separately. There are various ways to we can go about it.

My main point, though, is that it should be more apparent what build inclusion brings the cycle. If you print only the cycle nodes, it's not apparent.

> The https://github.com/gradle/gradle/pull/31070, stacked on the current, is improving the cycle detection logic, and guarantees, that cycle path will always be started from "acyclic" node, what is a plugin build node in case of build tree.

Again, it's strange to review something here that we are changing right away. The new PRs should mostly be merged together.

> always be started from "acyclic" node, what is a plugin build node in case of build tree.

But the cycle does not have to _start_ with a plugin build, right? It can start with a library build, and as long as there is a plugin build _in_ the cycle, we don't allow it. I think we should make it clear which invariant we are guarding against.

Reply:
> My main point, though, is that it should be more apparent what build inclusion brings the cycle

I got your point, but how is full path printed helps with that? 
Full path:
`: -> :a -> :b -> :c -> :d -> :e -> :c`
Answering at your question "what build inclusion brings the cycle", what we can say from the output above? `:b -> :c` ? But why not `:e -> :c`? Or `:d -> :e`?

Reply:
> A good parallel is a StackOverlow exception

For me, it's a good analogy, but that proves opposite thing - if you know, which recursive function is leading to SOE, the upstream to that function is not really important. What is important is _why_ the recursion isn't terminating.

Same for our case. Having additional `: -> :a -> :b ->` to the reported `:c -> :d -> :e -> :c` is doesn't help much, because the cycle enclosed not at `:b -> :c` edge, but at `:e -> :c`, which is reported originally.

Reply:
Yes, as I said, I don't feel strongly about this, we can leave the error message as is